### PR TITLE
llvmPackages_*.clangUseLLVM: add libunwind to lib search path (#201591)

### LINKED ISSUE #201591: llvmPackages.clangUseLLVM cannot find libunwind.so
### Describe the bug
When using `llvmPackages.clangUseLLVM` to compile, the linker cannot find libunwind.so.

### Steps To Reproduce
```
$ echo "int main() { }" > a.c
$ nix shell nixpkgs#llvmPackages.clangUseLLVM -c clang /tmp/a.c
/nix/store/039g378vc3pc3dvi9dzdlrd0i4q93qwf-binutils-2.39/bin/ld: cannot find -l:libunwind.so: No such file or directory
clang-11: error: linker command failed with exit code 1 (use -v to see invocation)
```

### Expected behavior
The linker should be able to find libunwind and compile successfully.

### Additional context
Run with `NIX_DEBUG=1` gives the following:
```
$ NIX_DEBUG=1 nix shell nixpkgs#llvmPackages.clangUseLLVM -c clang a.c
HARDENING: disabled flags: pie pic format stackprotector fortify strictoverflow
extra flags before to /nix/store/39yzdj9a7l8aw317l8la4kp6aalczsq7-clang-11.1.0/bin/clang:
  -Wl\,-dynamic-linker=/nix/store/4nlgxhb09sdr51nc9hdm8az5b08vzkgx-glibc-2.35-163/lib/ld-linux-x86-64.so.2
  -target
  x86_64-unknown-linux-gnu
original flags to /nix/store/39yzdj9a7l8aw317l8la4kp6aalczsq7-clang-11.1.0/bin/clang:
  a.c
extra flags after to /nix/store/39yzdj9a7l8aw317l8la4kp6aalczsq7-clang-11.1.0/bin/clang:
  -B/nix/store/4nlgxhb09sdr51nc9hdm8az5b08vzkgx-glibc-2.35-163/lib/
  -idirafter
  /nix/store/4pqv2mwdn88h7xvsm7a5zplrd8sxzvw0-glibc-2.35-163-dev/include
  --gcc-toolchain=/nix/store/1gf2flfqnpqbr1b4p4qz2f72y42bs56r-gcc-11.3.0
  -B/nix/store/fbc6m1s0638pf3vk5wgmcsbap8f04nar-clang-11.1.0-lib/lib
  -rtlib=compiler-rt
  -Wno-unused-command-line-argument
  -B/nix/store/lkckixdqkvwscdr5jf0v07w77lhlm70f-compiler-rt-libc-11.1.0/lib
  --unwindlib=libunwind
  -resource-dir=/nix/store/6bmyzky9i7rd2dp3lxc47rlabrhcl46y-clang-wrapper-11.1.0/resource-root
  -B
  /nix/store/6bmyzky9i7rd2dp3lxc47rlabrhcl46y-clang-wrapper-11.1.0/resource-root/lib
  -B/nix/store/6bmyzky9i7rd2dp3lxc47rlabrhcl46y-clang-wrapper-11.1.0/bin/
  -L/nix/store/4nlgxhb09sdr51nc9hdm8az5b08vzkgx-glibc-2.35-163/lib
  -L/nix/store/fbc6m1s0638pf3vk5wgmcsbap8f04nar-clang-11.1.0-lib/lib
HARDENING: disabled flags: relro pie bindnow
extra flags before to /nix/store/039g378vc3pc3dvi9dzdlrd0i4q93qwf-binutils-2.39/bin/ld:
  ''
original flags to /nix/store/039g378vc3pc3dvi9dzdlrd0i4q93qwf-binutils-2.39/bin/ld:
  --eh-frame-hdr
  -m
  elf_x86_64
  -o
  a.out
  /nix/store/4nlgxhb09sdr51nc9hdm8az5b08vzkgx-glibc-2.35-163/lib/crt1.o
  /nix/store/4nlgxhb09sdr51nc9hdm8az5b08vzkgx-glibc-2.35-163/lib/crti.o
  /nix/store/6bmyzky9i7rd2dp3lxc47rlabrhcl46y-clang-wrapper-11.1.0/resource-root/lib/linux/clang_rt.crtbegin-x86_64.o
  -L/nix/store/4nlgxhb09sdr51nc9hdm8az5b08vzkgx-glibc-2.35-163/lib
  -L/nix/store/fbc6m1s0638pf3vk5wgmcsbap8f04nar-clang-11.1.0-lib/lib
  -L/nix/store/1gf2flfqnpqbr1b4p4qz2f72y42bs56r-gcc-11.3.0/lib64/gcc/x86_64-unknown-linux-gnu/11.3.0
  -L/nix/store/1gf2flfqnpqbr1b4p4qz2f72y42bs56r-gcc-11.3.0/lib64/gcc/x86_64-unknown-linux-gnu/11.3.0/../../../../lib64
  -L/nix/store/1gf2flfqnpqbr1b4p4qz2f72y42bs56r-gcc-11.3.0/lib64/gcc/x86_64-unknown-linux-gnu/11.3.0/../../..
  -dynamic-linker=/nix/store/4nlgxhb09sdr51nc9hdm8az5b08vzkgx-glibc-2.35-163/lib/ld-linux-x86-64.so.2
  /tmp/a-804f0d.o
  /nix/store/6bmyzky9i7rd2dp3lxc47rlabrhcl46y-clang-wrapper-11.1.0/resource-root/lib/linux/libclang_rt.builtins-x86_64.a
  --as-needed
  -l:libunwind.so
  --no-as-needed
  -lc
  /nix/store/6bmyzky9i7rd2dp3lxc47rlabrhcl46y-clang-wrapper-11.1.0/resource-root/lib/linux/libclang_rt.builtins-x86_64.a
  --as-needed
  -l:libunwind.so
  --no-as-needed
  /nix/store/6bmyzky9i7rd2dp3lxc47rlabrhcl46y-clang-wrapper-11.1.0/resource-root/lib/linux/clang_rt.crtend-x86_64.o
  /nix/store/4nlgxhb09sdr51nc9hdm8az5b08vzkgx-glibc-2.35-163/lib/crtn.o
extra flags after to /nix/store/039g378vc3pc3dvi9dzdlrd0i4q93qwf-binutils-2.39/bin/ld:
  -rpath
  /nix/store/4nlgxhb09sdr51nc9hdm8az5b08vzkgx-glibc-2.35-163/lib
/nix/store/039g378vc3pc3dvi9dzdlrd0i4q93qwf-binutils-2.39/bin/ld: cannot find -l:libunwind.so: No such file or directory
clang-11: error: linker command failed with exit code 1 (use -v to see invocation)
```

It is clear that the path of libunwind is not added to search path. Maybe we should add `-L${targetLlvmLibraries.libunwind}/lib` in `nixSupport.cc-cflags`.

https://github.com/NixOS/nixpkgs/blob/3a86856a13c88c8c64ea32082a851fefc79aa700/pkgs/development/compilers/llvm/git/default.nix#L151-L172

### Notify maintainers
@Ericson2314 @primeos

### Metadata
- system: `"x86_64-linux"`
 - host os: `Linux 6.0.8, NixOS, 22.11 (Raccoon), 22.11pre-git`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.11.0`
 - channels(root): `"nixos-22.05"`
 - channels(sharzy): `"nixos, nixpkgs"`
 - nixpkgs: `/home/sharzy/.nix-defexpr/channels/nixpkgs`

### PR DESCRIPTION
**Pull request by SharzyL**
###### Description of changes

resolves #201591

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by sternenseemann** [comment]
I guess that brings us to the old question of whether `clangUseLLVM` should be used/usable for anything else but as an implementation detail of how `pkgsLLVM.stdenv` is bootstrapped. But since it is exposed so prominently it should probably do something, @Ericson2314?

**Comment by ehmry** [comment]
Stale but related #207325.

File: pkgs/development/compilers/llvm/10/default.nix
Comment (pwaller) [inline pkgs/development/compilers/llvm/10/default.nix:158]: Given that the library is conditional on !isWasm below and targetPlatform.useLLVM, this should probably be along side that.

**Review by pwaller — COMMENTED** [review]
I think the change is needed to make `nix shell nixpkgs#pkgsLLVM.stdenv.cc` be a functioning compiler.

On master prior to this patch, clangUseLLVM gives you a functioning compiler if the stdenv is used to build software, but then if you try to use the compiler outside of the stdenv it fails. The reason is that the binutils setup hook propagates libunwind (and other runtimes) into the `NIX_LDFLAGS` when you're in the stdenv because these packages are specified in `extraPackages`.

Since `-lunwind` and friends are a part of the compiler wrapper, it makes sense to me that the `-L` flags should also be a part of the wrapper, as you have done here.

Just one comment that they'll need conditionalising, I think `-L/path/to/unwind` should only be added if `-lunwind` is being added, it might be that the package won't build outside of those scenarios for example.

**Comment by pwaller** [comment]
cc @reckenrode are you aware of whether this could cause problems on darwin?

**Comment by reckenrode** [comment]
> cc @reckenrode are you aware of whether this could cause problems on darwin?

I shouldn’t think so. The only risk would be if it could cause an infinite recursion, but the ofborg eval check should catch that.

File: pkgs/development/compilers/llvm/10/default.nix
Comment (SharzyL) [reply pkgs/development/compilers/llvm/10/default.nix:158]: `clang` links against libunwind when `-unwindlib=libunwind` is set. Hence if conditioning on `targetPlatform.useLLVM`, `llvmPackages.clangUseLLVM` still cannot find libunwind.

**Comment by SharzyL** [comment]
Rebased the commit and added conditioning on `!isWasm` now.

**Review by pwaller — APPROVED** [review]
Seems reasonable to me, unfortunately I don't have commit rights. @reckenrode ? @RaitoBezarius ?

File: pkgs/development/compilers/llvm/10/default.nix
Comment (rrbutani) [reply pkgs/development/compilers/llvm/10/default.nix]: It's definitely curious that we gate `-lunwind` on `targetPlatform.useLLVM` but not `--unwindlib=libunwind`.

Tracing back to [where the `-lunwind` was added](https://github.com/NixOS/nixpkgs/commit/37194a325de090747ce5f97a38de4329708c1912#diff-5c97508ac7912c7688e27a77c3ddca9f3c8e5a5fc448d79736561b771ca64cebR125-R126) doesn't reveal any obvious reason for the discrepancy and I'm actually not clear on why the `-lunwind` was needed in the first place (my understanding is that [it's implied](https://github.com/llvm/llvm-project/blob/a78f5c064975f4d597ed1b18206d579748eb91cb/clang/lib/Driver/ToolChains/CommonArgs.cpp#L1887)). Perhaps there were scenarios where the linker was being invoked directly with the expectation that libunwind would be pulled in?

Regardless, doesn't need to block this PR. Adding this `-L` in shouldn't hurt.

**Review by rrbutani — APPROVED** [review]
I think @sternenseemann's question is the right one. It would be nicer to have something other than `clangUseLLVM` intended for use outside of `stdenv` but in the interim I think it makes sense to merge this fix.

It would be nice if we could have a test guarding against regressions but that doesn't need to block this PR.

@SharzyL thanks for your patience.

**Comment by RaitoBezarius** [comment]
Apologies for the delay, @SharzyL can you rebase for the conflicts and after that we go for the merge IMHO.

**Comment by SharzyL** [comment]
Rebase is complete

**Comment by RaitoBezarius** [comment]
This probably should not have targeted staging given the low amount of rebuilds, but well, it can wait staging :).

**Comment by wegank** [comment]
Is the exclusion of LLVM 17 intended in this PR?

**Comment by SharzyL** [comment]
Probably a omittance during rebase

**Pull request merged**
